instrumentation/opentelemetry-instrumentation-asgi/src/opentelemetry/instrumentation/asgi/__init__.py

```diff
@@ -729,7 +725,20 @@ async def __call__(
                 )
 
                 await self.app(scope, otel_receive, otel_send)
+        except Exception as exc:  # pylint: disable=broad-except
+            exception = exc
         finally:
+            if exception is not None and _report_new(
+                self._sem_conv_opt_in_mode
+            ):
+                _set_status(
+                    span=span,
+                    metrics_attributes=attributes,
+                    status_code=-1,
```

Comment:
The `-1` status code is a bit awkward. The current change is fine but I would like to brainstorm some ways we can abstract all of this logic out somehow. The following are probably required in some way or another and should be shared across all instrumentations (especially wsgi and asgi) : 1. Setting status code on span 2. Setting error attributes on span if recording and if status code is erroneous 3. Setting error attributes on metrics attributes if status code is erroneous 4. Setting error attributes on span/metrics in general.

Reply:
I wouldn't say I like it either. We are actually doing that in several places where we try to int the status_code, and if it fails, we set status_code as -1. Let me see what can I do here to improve the way we are doing this.